PRCS/prc-4.md

```diff
@@ -0,0 +1,139 @@
+    * Function to create a sell order. A sell order persists and has the following properties: 
+      * `assetAmount` - amount of the game asset the seller is selling,
+      * `price` - the price in native gas tokens of the base chain the seller is requesting,
+      * `seller` - the address of the seller,
+      * `active` - signalizes if the sell order is active or not (meaning it has been cancelled or filled)
+    * Function to cancel a sell order.
+    * Function to fill a sell order (in other words - buy). There is no "buy order" that persists, and rather the buy function directly transfers the value specified by price defined in the orders to the sellers. Buying marks all the sell orders it fulfills as inactive by changing the value of `active` flag to `false`.
+2. The game chain monitors this contract using CDEs and exposes an API to query its state.
+
+That is to say, the contract on the base chain has no way of knowing if somebody who makes a sell order really has that amount of game assets in their account. Rather,
+
+1. When somebody creates a sell order: The game chain sees it and checks if the user has enough of the asset in their account.
+    * If yes, the amount is locked so they cannot spend it in-game (unless they cancel the order).
+    * If not, the order is marked as invalid (to differentiate it from the case where it doesn't know the order exists yet).
+2. When somebody wants to buy, they specify the orders they wish to purchase by ID. 
+    * The responsibility of querying the API of game chain to check the validity of the sell order is left up to the front-end providers, as well as the presentation of the possible sell orders.
+3. The user then makes a smart contract call that fulfills all the orders and sends the right amount to the corresponding accounts in a single transaction (atomic).
+4. The game chain monitors the successfully fulfilled orders on the base chain and transfers the assets between accounts.
+
+**Note:**
+All actions (sell, buy, cancel) are done on the base chain (and not on the game chain). This allows the base chain to be the source of truth for the state of which orders are valid which avoids double-spends. That is to say, if you try and perform a buy order, you are guaranteed by the base chain itself that the order hasn't been fulfilled by anybody else yet.
+
+## Reference Implementation
+
+You can find all the contracts and interfaces in the Paima Engine codebase [here](https://github.com/PaimaStudios/paima-engine/blob/master/packages/contracts/evm-contracts/contracts/orderbook/).
+
+## Considerations
+
+### Avoiding many failed txs in this model
+
+The problem with this model is people will naturally all try to buy the order with the most favorable sell price. Chosen base chain should have fast blocks so collisions are not so likely if the UI updates fast enough, but if bots start trading good orders as soon as they appear it may cause a rush with many failed transactions.
```

Comment:
```suggestion
The problem with this model is people will naturally all try to buy the order with the most favorable sell price. The chosen base chain should have fast blocks so that collisions are not too likely if the UI updates fast enough, but if bots start trading good orders as soon as they appear it may cause a rush with many failed transactions.
```